sdk/core/core/src/az_http_response.c

```diff
@@ -18,7 +19,21 @@ AZ_NODISCARD AZ_INLINE az_result _az_is_char(az_span slice, uint8_t c)
   return az_span_ptr(slice)[0] == c ? AZ_OK : AZ_CONTINUE;
 }
 
-static AZ_NODISCARD az_result _az_is_a_colon(az_span slice) { return _az_is_char(slice, ':'); }
+static AZ_NODISCARD az_result _az_valid_header_name_to_colon(az_span slice)
+{
+  az_result is_colon_result = _az_is_char(slice, ':');
+  if (is_colon_result == AZ_OK)
+  {
+    return is_colon_result;
+  }
```

Comment:
This has a lot of ceremony. Also, I think we end up allowing an empty header name and also access a pointer value out-of-bounds. I don't think that is valid. We should do the validation check before this.

A test like this behaves incorrectly with `az_http_response_get_next_header` returning `AZ_OK`:
```C
(void)state;
  {
    az_http_response response = { 0 };
    assert_return_code(
        az_http_response_init(
            &response,
            AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n"
                             ": Value11\r\n"
                             "\r\n"
                             "KKKKKJJJJJIIIIIHHHHHGGGGGFFFFFEEEEEDDDDDCCCCCBBBBBAAAAA")),
        AZ_OK);

    az_http_response_status_line status_line = { 0 };
    assert_return_code(az_http_response_get_status_line(&response, &status_line), AZ_OK);
    az_pair header = { 0 };
    az_result fail_header_result = az_http_response_get_next_header(&response, &header);
    assert_true(AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER == fail_header_result);
  }
```

The `_az_span_scan_until` is doing a lot of unnecessary work, especially for what we are trying to do. We may want to get rid of and just use find.

Reply:
`remaining` in `az_http_response_get_next_header` could be of size 0 here, and now we are reading from memory we shouldn't have access to:
https://github.com/Azure/azure-sdk-for-c/blob/de74386a0dc9051ba16229b705ad574fc1069a26/sdk/core/core/src/az_http_response.c#L171

Reply:
it can't be size 0 after reading a header.
RFC asks /r/n after all headers and before a body.
So, after reading a header that ends in `\r\n` we can have only two options
- Another header starts
or
- `\r\n` again as the end of headers

Reply:
What is preventing the user providing incomplete or invalid response, or one that doesn't have any headers at all? It is just buffer with data that our API accepts and tries to parse. 

During that, we are reading memory that we shouldn't. That is not a good idea, in almost any circumstance.

Here's a test case:
```C

static void test_http_response_header_validation_fail(void** state)
{
  (void)state;
  {
    az_http_response response = { 0 };
    assert_return_code(
        az_http_response_init(
            &response,
            AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n")
            ),
        AZ_OK);

    az_http_response_status_line status_line = { 0 };
    assert_return_code(az_http_response_get_status_line(&response, &status_line), AZ_OK);
    az_pair header = { 0 };

    // This is reading from a memory address it shouldn't.
    az_result fail_header_result = az_http_response_get_next_header(&response, &header);
    assert_int_equal(AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER, fail_header_result);
  }
}
```